# Bustamante Attack Solution with 
There are two approach to mitigate Bustamante attack included in this project. One is using Sparse Vector Technique and another one is randomized response with biasing. Both of the technique are scripted into python files 'vcf_bit_sparse_vector.py' and 'vcf_bit_rand_response.py' respectively.

Based on Utility (accuracy in output) and attack mitigation randomized response performs well. With bias >=2 its shown that the attack is not present for the given vcf files of population 500. There can be higher bias possible as well as this parameter is experimental. Higher bias will bring more incorrect result in the beacon service. Even with bias=2, we can achieve 99% sensitivity and specificity and also mitigate the attack. Some results are given in 'results.txt'

Some pkl files are available in 'pkl_files' folder. The data is from https://app.box.com/files/0/f/11133431279/wholeRecordsData.

**The solution is in python 3 (sorry I am new in python), add 'alias python=python3' if you face error!**

>python vcf_bit_rand_response.py -f vcfSize500/vcfSize500/chr10_selectedInd.vcf -n 500 -b 50

Input: a)VCF file as -f  (ie, vcfSize500/vcfSize500/chr10_selectedInd.vcf)
			b)Population size in -n (ie, 500)
			c)Bias of the randomized response (-b), the less the bias the more the accurate the result is For example, if bias = 5 the output is truthful if the random probability (taken from 0-100) is > 5 .
	
Output: one pkl file with the bias given as input
	
>python vcf_bit_sparse_vector.py -f vcfSize500/vcfSize500/chr10_selectedInd.vcf -n 500 -e .5 -r chr10returnValue_delta_1.0.txt

Input: a)VCF file as -f  (ie, vcfSize500/vcfSize500/chr10_selectedInd.vcf)
			b)Population size in -n (ie, 500)
			c)Epsilon -e (ie, .5, 1, 2)
			d)Output from build_beacon_delta_rare.py file in rare file -r (ie, chr10returnValue_delta_1.0.txt where delta is 1.0)
	
Output: one pkl file with the epsilon given as input

>python utility_test.py -f pkl_files/vcfSize500wholeRecords.pkl -n pkl_files/vcfSize500WholeRecords_rand_res_bias_2_both.pkl

Input: 
	a) main pkl file which contains the original results in bitarray (ie, pkl_files/vcfSize500wholeRecords.pkl) 
	b)solution pkl file generated by 'vcf_bit_sparse_vector.py' or 'vcf_bit_rand_response.py' (ie, pkl_files/vcfSize500WholeRecords_rand_res_bias_2_both.pkl)
	
Output: true_positive, true_negetive, false_positive, false_negetive, sensitivity, specificity of the results
	

>python build_beacon_delta_rare.py -t 1 -d 1 -f vcfSize500/vcfSize500/chr10_selectedInd.vcf

this script is to simulate beacon database construction (changed from original)
	
>python build_beacon_delta_rare.py -t returnThreshold -d errorRate -f vcf_file_location

return threshold T(1, 2): beacon return "true" when at least t individuals in beacon have query allele
error rate D: mask D(1e-06 ~ 1.0) rare SNPs, force to return "false" (1e-06 is the empirical sequencing error rate, which indicates no error introduced to database, except sequencing error)
 input: training dataset vcf file(vcfSize100.tar.gz/vcfSize200.tar.gz/vcfSize500.tar.gz)
output: beacon return result & allele freq, this file is required for sparse vector technique

>python Query_fromBeacon_delta_wholeRecords.py -n populationSize -t returnThreshold -d delta -f1 solution_pkl_file -f2 not_in_dataset_file -r rare_file_from_build_beacon_delta_rare.py

this script is to perform allele frequency attack based on log likelihood ratio test(lrt)
-n population size
-return threshold T(1, 2): beacon return "true" when at least t individuals in beacon have query allele
-error rate D: error rate(1e-06 ~ 1.0) on rare SNPs (used as a parameter in Bustamante attack), D rare SNPs masked in Beacon database (return "False") 
	
[PLEASE NOTE: parameters T = 1, D = 1e-06 should be set to ensure the correct result with the input examples, if you would like to try different parameters, please re-run build_beacon_delta_rare.py with the corresponding parameters to obtain the correct input file of beacon return result & allele freq] 
	input:  case(not in beacon) & test(in beacon) individual variation info (500notInBeaconwholeRecords.pkl, vcfSize500wholeRecords.pkl)               
	        beacon return result & allele freq (chr10returnValue_500wholeR.txt)  
	-f1 and f2 are the pkl files (same as previous python file)
	-r output from build_beacon_delta_rare.py file with that delta and threshold
	
> python power_lrt.py -d delta -n population -f lrt_file_from_Query_fromBeacon_delta_wholeRecords.py

this script is to calculate attack power based on lrt scores (display is turned off for ubuntu server edition)
	input: lrt scores from previous step (need to specify input files in script)(chr10_lrtScoreByStep_delta_x.txt)
	output: figure w/ specified inputs (e.g. power of different error rates)
	

![alt text][logo]

[logo]: http://www.cs.umanitoba.ca/~bruce/umanlogotight.jpg "University of Manitoba"
	
